metaflow_extensions/plugins/pip_step_decorator.py

```diff
@@ -0,0 +1,150 @@
+
+    def task_pre_step(
+        self,
+        step_name,
+        task_datastore,
+        metadata,
+        run_id,
+        task_id,
+        flow,
+        graph,
+        retry_count,
+        max_user_code_retries,
+        ubf_context,
+        inputs,
+    ):
+        """Install packages with pip."""
+        flow_dir = Path(sys.argv[0]).parent
+        path = (flow_dir / self.attributes["path"]) if self.attributes["path"] else None
+        libraries = self.attributes["libraries"]
+
+        if path is not None:
+            pip_install_reqs(path)
+
+        if libraries is not None:
+            pip_install_libraries(libraries)
+
+    def task_post_step(
+        self, step_name, flow, graph, retry_count, max_user_code_retries
+    ):
+        """After step has run, ensure local conda environment is fresh."""
+        ensure_conda_integrity(step_name, flow, graph)
+        return
+
+    def task_exception(
+        self, exception, step_name, flow, graph, retry_count, max_user_code_retries
+    ):
+        """After step exception, ensure local conda environment is fresh."""
+        ensure_conda_integrity(step_name, flow, graph)
+        return
+
+
+def is_task_local() -> bool:
+    """True if process is running on the same machine as the local orchestrator."""
+    return not any(f"--with {x}" in " ".join(sys.argv) for x in REMOTE_RUNTIMES)
+
+
+def ensure_conda_integrity(step_name: str, flow: FlowSpec, graph: FlowGraph) -> None:
+    """If using conda and local runtime, ensure subsequent steps get clean env."""
+    if is_task_local() and is_mflow_conda_environment():
+        step_decorators = graph.nodes[step_name].decorators
+        # Triggers re-creation of Conda env after extra installation actions
+        # has possibly 'polluted' it.
+        clear_local_conda_cache(flow.name, step_decorators)
+    return
+
+
+def clear_local_conda_cache(
+    flow_name: str, step_decorators: List[StepDecorator]
+) -> None:
+    """Clear `.metaflow/{flow_name}/conda.dependencies`."""
+    conda_decorator = next(
+        d for d in step_decorators if isinstance(d, CondaStepDecorator)
+    )  # There can be only one Conda decorator per step
+
+    local_root = conda_decorator.local_root  # Local .metaflow folder
+    conda_deps_path = Path(local_root) / flow_name / "conda.dependencies"
+    # Existence check for robustness, this method could get called multiple times
+    # (in post `task_post_step` and `task_exception`) depending on retry behaviour
+    if conda_deps_path.exists():
+        conda_deps_path.unlink()
+
+    return
+
+
+def fill_constraint(v: str) -> str:
+    """Add constraint `v` if not already prefixed with equality constraint."""
+    # https://www.python.org/dev/peps/pep-0440/#version-specifiers
+    constraints = {"==", ">=", "<=", "<", ">", "~=", "!="}
+
+    if v == "":
+        return ""
+    elif not any(v.startswith(str_op) for str_op in constraints):
+        return "==" + v
+    else:
+        return v
```

Comment:
>    v to a more semantic name --> pkg_version?

Not pedantic, completely reasonable - just one of those things that didn't get renamed after the initial test function

>    hard-coded variable to global

Fair. I can never decide when to make something a global constant or not - if it's only used in one place then I tend to encapsulate it within the function so no-one is tempted to override the global but this is probably paranoia.

> can compactify the logic whilst maintaining legibility

* I prefer the use of `map` but have recently switched to using comprehensions when possible because it's apparently easier for non-functional people to get - thoughts?
* Agree the `v==""` is redundant (captured by the `else` clause) but I prefer to avoid mutation of `pkg_version`